llvm/lib/Analysis/BasicAliasAnalysis.cpp

```diff
@@ -417,7 +417,8 @@ static LinearExpression GetLinearExpression(
                                 Depth + 1, AC, DT);
         E.Offset <<= RHS.getLimitedValue();
         E.Scale <<= RHS.getLimitedValue();
-        E.IsNSW &= NSW;
+        // The nsw flag has different semantics for shift and mul.
+        E.IsNSW = false;
```

Comment:
This is conceptually right, but you should keep the nsw flag if you know that either RHS is a constant other than bitwidth-1 or the LHS is a constant other than -1.

Please split off this part of the patch with a minimal test that demonstrates that the current behavior is wrong in this specific edge case.

Reply:
Well, I guess LHS being a constant isn't relevant in this code, so just checking the RHS is enough.

Reply:
Sorry for my misunderstanding. `shl nsw %a, c1` is equivalent to `mul nsw %a, (1<<c1)`.
Alive2: https://alive2.llvm.org/ce/z/clpceU
I have found the root cause of this issue. I will propose a fix soon.